go/client/cmd_send.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2015 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package client
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+
+	"golang.org/x/net/context"
+
+	"github.com/keybase/cli"
+	"github.com/keybase/client/go/libcmdline"
+	"github.com/keybase/client/go/libkb"
+	keybase1 "github.com/keybase/client/go/protocol"
+	"github.com/keybase/gregor/protocol/chat1"
+	"github.com/keybase/gregor/protocol/gregor1"
+)
+
+type cmdSend struct {
+	libkb.Contextified
+	tlfName string
+	message string
+}
+
+func newCmdSend(cl *libcmdline.CommandLine, g *libkb.GlobalContext) cli.Command {
+	return cli.Command{
+		Name:         "send",
+		Usage:        "Send a message to a conversation",
+		ArgumentHelp: "<conversation> <message>",
+		Action: func(c *cli.Context) {
+			cl.ChooseCommand(&cmdSend{Contextified: libkb.NewContextified(g)}, "send", c)
+		},
+	}
+}
+
+// getConversationID returns the most recent conversation's ConversationID for
+// given TLF ID, or creates a new conversation and returns its ID if none
+// exists yet.
+//
+// TODO: after we implement multiple conversations per TLF and topic names,
+// replace this with something that looks up by topic name
+func (c *cmdSend) getConversationID(ctx context.Context, tlfID chat1.TLFID, chatClient keybase1.ChatLocalClient) (id chat1.ConversationID, err error) {
+	ipagination := &chat1.Pagination{}
+getinbox:
+	for {
+		ipagination.Num = 32
+		iview, err := chatClient.GetInboxLocal(context.TODO(), ipagination)
+		if err != nil {
+			return id, err
+		}
+		for _, conv := range iview.Conversations {
+			if bytes.Equal(conv.Metadata.IdTriple.Tlfid, tlfID) {
+				return conv.Metadata.ConversationID, nil
+			}
+		}
+
+		if iview.Pagination == nil || iview.Pagination.Last != 0 {
+			break getinbox
+		} else {
+			ipagination = iview.Pagination
+		}
+	}
+
+	id, err = chatClient.NewConversationLocal(ctx, chat1.ConversationIDTriple{
+		Tlfid:     tlfID,
+		TopicType: chat1.TopicType_CHAT,
+		// TopicID filled by server?
+	})
+	if err != nil {
+		return id, err
+	}
+
+	return id, nil
+}
+
+func (c *cmdSend) Run() (err error) {
+	uid := c.G().Env.GetUID()
+	if uid.IsNil() {
```

Comment:
what happens if the user isn't logged in -- this?  Maybe we need a slightly nicer experience then.